hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -372,12 +374,12 @@ def _repr_step(self, rule, data, result):
         return f"{output_assignment}state.{rule.function.__name__}({args})"
 
     def _add_result_to_targets(self, targets, result):
-        name = self._new_name()
-        self.__printer.singleton_pprinters.setdefault(
-            id(result), lambda obj, p, cycle: p.text(name)
-        )
-        self.names_to_values[name] = result
         for target in targets:
+            name = self._new_name(target)
+            self.__printer.singleton_pprinters.setdefault(
+                id(result), lambda obj, p, cycle: p.text(name)
+            )
```

Comment:
[The lint rule](https://docs.astral.sh/ruff/rules/function-uses-loop-variable/) is pointing out that every single lambda you're defining here will print the name defined in the _last_ loop iteration.  [I implemented that](https://github.com/PyCQA/flake8-bugbear/pull/265) because it tends to catch real bugs!

As a result, I'm also nervous because the tests are passing, and so we must not have anthing that checks for correct printing of bundle names from a rule which returns multiple targets.  Can you add a test for this?  The lambda is easy to fix with a binding `lambda obj, p, cycle, *, name=name: p.text(name)`.